File: examples/java-junit-gradle/src/test/java/examples/Library.java
package examples;

import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.List;

public final class Library {

    public static final int FEE_PENCE_PER_DAY = 50;

    public record Loan(String title, LocalDate due) {}

    private Library() {}

    /** Fee in pence for returning a loan: 50p per day past the due date. */
    public static int lateFee(Loan loan, LocalDate returnedOn) {
        long daysLate = Math.max(0, ChronoUnit.DAYS.between(loan.due(), returnedOn));
        return Math.toIntExact(daysLate) * FEE_PENCE_PER_DAY;
    }

    /** A member may borrow as long as none of their loans is overdue. */
    public static boolean mayBorrow(List<Loan> loans, LocalDate on) {
        return loans.stream().noneMatch(loan -> loan.due().isBefore(on));
    }
}

File: examples/kotlin-kotest/src/test/kotlin/examples/Library.kt
package examples

import java.time.LocalDate
import java.time.temporal.ChronoUnit

const val FEE_PENCE_PER_DAY = 50

data class Loan(val title: String, val due: LocalDate)

/** Fee in pence for returning a loan: 50p per day past the due date. */
fun lateFee(loan: Loan, returnedOn: LocalDate): Int {
    val daysLate = maxOf(0, ChronoUnit.DAYS.between(loan.due, returnedOn))
    return daysLate.toInt() * FEE_PENCE_PER_DAY
}

/** A member may borrow as long as none of their loans is overdue. */
fun mayBorrow(loans: List<Loan>, on: LocalDate): Boolean = loans.none { it.due.isBefore(on) }
